rollout/trafficrouting/traefik/traefik.go

```diff
@@ -0,0 +1,164 @@
+}
+
+type ClientInterface interface {
+	Get(ctx context.Context, name string, options metav1.GetOptions, subresources ...string) (*unstructured.Unstructured, error)
+	Update(ctx context.Context, obj *unstructured.Unstructured, options metav1.UpdateOptions, subresources ...string) (*unstructured.Unstructured, error)
+}
+
+func NewReconciler(cfg ReconcilerConfig) *Reconciler {
+	reconciler := &Reconciler{
+		Rollout: cfg.Rollout,
+		Client:  cfg.Client,
+	}
+	return reconciler
+}
+
+func NewDynamicClient(di dynamic.Interface, namespace string) dynamic.ResourceInterface {
+	return di.Resource(GetMappingGVR()).Namespace(namespace)
+}
+
+func GetMappingGVR() schema.GroupVersionResource {
+	group := defaults.DefaultTraefikAPIGroup
+	parts := strings.Split(defaults.DefaultTraefikVersion, "/")
+	version := parts[len(parts)-1]
+	resourceName := apiGroupToResource[group]
+	return schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resourceName,
+	}
+}
+
+func (r *Reconciler) UpdateHash(canaryHash, stableHash string, additionalDestinations ...v1alpha1.WeightDestination) error {
+	return nil
+}
+
+func (r *Reconciler) SetWeight(desiredWeight int32, additionalDestinations ...v1alpha1.WeightDestination) error {
+	ctx := context.TODO()
+	rollout := r.Rollout
+	traefikServiceName := rollout.Spec.Strategy.Canary.TrafficRouting.Traefik.TraefikServiceName
+	traefikService, err := r.Client.Get(ctx, traefikServiceName, metav1.GetOptions{})
+	if err != nil || traefikService == nil {
+		return err
+	}
+	canaryServiceName := rollout.Spec.Strategy.Canary.CanaryService
+	stableServiceName := rollout.Spec.Strategy.Canary.StableService
+	services, isFound, err := unstructured.NestedSlice(traefikService.Object, "spec", "weighted", "services")
+	if err != nil || !isFound {
+		return err
+	}
+	canaryService, err := getService(canaryServiceName, services)
+	if err != nil || canaryService == nil {
+		return err
+	}
+	err = unstructured.SetNestedField(canaryService, int64(desiredWeight), "weight")
+	if err != nil {
+		return err
+	}
+	stableService, err := getService(stableServiceName, services)
+	if err != nil || stableService == nil {
+		return err
+	}
+	err = unstructured.SetNestedField(stableService, int64(100-desiredWeight), "weight")
+	if err != nil {
+		return err
+	}
+	err = unstructured.SetNestedSlice(traefikService.Object, services, "spec", "weighted", "services")
+	if err != nil {
+		return err
+	}
+	_, err = r.Client.Update(ctx, traefikService, metav1.UpdateOptions{})
+	return err
+}
+
+func getService(serviceName string, services []interface{}) (map[string]interface{}, error) {
+	var selectedService map[string]interface{}
+	for _, service := range services {
+		typedService, ok := service.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("Failed type assertion setting weight for traefik service")
+		}
+		nameOfCurrentService, isFound, err := unstructured.NestedString(typedService, "name")
+		if err != nil || !isFound {
```

Comment:
if `!isFound` it will return a nil error. Better handle this case separately and return proper error messages.